File: Leetcode/Minimum Difficulty of a Job Schedule/Minimum Difficulty of a Job Schedule.py
# You want to schedule a list of jobs in d days. Jobs are dependent (i.e To work on the i-th job, you have to finish all the jobs j where
#  0 <= j < i).
# You have to finish at least one task every day. The difficulty of a job schedule is the sum of difficulties of each day of the d days. 
# The difficulty of a day is the maximum difficulty of a job done in that day.
# Given an array of integers jobDifficulty and an integer d. The difficulty of the i-th job is jobDifficulty[i].
# Return the minimum difficulty of a job schedule. If you cannot find a schedule for the jobs return -1.



class Solution:
    def minDifficulty(self, jobDifficulty: List[int], d: int) -> int:
        if d > len(jobDifficulty):
            return -1
        
        # index starts from zero.
        # d[i][j]: the minimum job difficulty after i-th cut,
        # range within the first j+1 jobs (job at index j)
        # totally need to make d-1 cuts, and range within all
        # of the len(jobDifficulty) jobs.
        
        dp = {}
        
        # initialization
        running_max = 0
        for k in range(len(jobDifficulty)):
            running_max = max(running_max, jobDifficulty[k])
            dp[(0, k)] = running_max

        # obtain dp[i][j]
        for i in range(1, d):
            # the i-th cut is at least after the i-th element,
            # since each day at least one job has to be scheduled.
            # all days can be included in the last day of the current
            # schedule, so j is in range(i+1, length_of_jobs)
            for j in range(i, len(jobDifficulty)):
                # to obtain dp[i][j], we need to find out all possibilities
                # to schedule i+1 jobs within the first j+1 jobs.
                # since we already know dp[i-1][k] for all k, it suffices
                # to take the minimum of all 
                # (dp[i-1][k] + max(jobDifficulty[k:j+1])) for i <= k <= j
                running_max = 0
                # running_min
                dp[(i,j)] = 2**16
                for k in range(j, i-1, -1):
                    running_max = max(running_max, jobDifficulty[k])
                    dp[(i, j)] = min(dp[(i, j)], dp[(i-1, k-1)] + running_max)
            
            
        return dp[(d-1, len(jobDifficulty)-1)]

File: Leetcode/Minimum Difficulty of a Job Schedule/Minimum Difficulty of a Job Schedule.java
class Solution {
    public int minDifficulty(int[] jobDifficulty, int d) {
        //Finding the size of array
        int n = jobDifficulty.length;
        //If size is less than d, thenm return -1
        if(n < d){
           return -1; 
        } 
        //Initialising dp array
        int[][] dp = new int[d][n];
        
        //Base case value will be equal to first value in the array
        dp[0][0] = jobDifficulty[0];
        //Looping through the difficulty array
        for(int i = 1; i < n; i++){
            //For value of d as 0, dp value will be max of difficulty and previous dp value for each array element
            dp[0][i] = Math.max(jobDifficulty[i],dp[0][i-1]);
        }
        
        //Looping through each d
        for(int i = 1; i < d; i++){
            //Looping through difficulty array
            for(int j = i; j < n; j++){
                //Defining local max as difficulty element
                int localMax = jobDifficulty[j];
                //Initialising dp value as max integer
                dp[i][j] = Integer.MAX_VALUE;
                //Looping from j value to i value, finding local maximum with respect to difficulty array on that index, and then finding dp value as minimum of current value and (dp[i-1][r-1] + localMax).
                for(int r = j; r >= i; r--){
                    localMax = Math.max(localMax,jobDifficulty[r]);
                    dp[i][j] =  Math.min(dp[i][j],dp[i-1][r-1] + localMax);
                }
            }
        }
        
        //Returning the ans
        return dp[d-1][n-1];
    }
}
